### ISSUE TITLE
W1b: `undistort_points_kannala_brandt` — the principal point returns `nan` in float16, because both singularity guards underflow

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4308]
`undistort_points_kannala_brandt` returns `nan` for a point **at the principal point** when `params` is `float16`.

The function guards its two singular spots with hardcoded constants:

```python
th = rth.clamp(min=1e-16).sqrt()   # Newton start
...
denom = rth + eps                  # eps = 1e-8, final radial rescale
mag = radius_undistorted.abs() / denom
```

Both constants underflow to exactly `0.0` in `float16` (`1e-8` is below the smallest float16 subnormal, `~5.96e-8`; `1e-16` is far below it). The whole body runs in `params.dtype`, so with `float16` `params` the guards are no guards at all: at the principal point `rth == 0`, `denom == 0 + 0 == 0`, `radius_undistorted == tan(0) == 0`, and the result is `0.0 / 0.0 == nan` in both components.

`float32`, `float64` and `bfloat16` all return the exact origin — `bfloat16` survives because it keeps the float32 exponent range, so the underflow is specific to `float16`. The sibling `undistort_points_affine` has no such constants and returns the origin in every dtype.

## Repro

```python
import torch
from kornia.geometry.camera.distortion_affine import undistort_points_affine
from kornia.geometry.camera.distortion_kannala_brandt import undistort_points_kannala_brandt

for dtype in (torch.float64, torch.float32, torch.bfloat16, torch.float16):
    params = torch.tensor([100.0, 100.0, 4.0, 3.0, 0.1, 0.01, 0.001, 0.0001], dtype=dtype)
    principal_point = torch.tensor([4.0, 3.0], dtype=dtype)   # == (cx, cy)
    print(dtype, undistort_points_kannala_brandt(principal_point, params).tolist())

for dtype in (torch.float32, torch.float16):                  # the same on mps
    params = torch.tensor([100.0, 100.0, 4.0, 3.0, 0.1, 0.01, 0.001, 0.0001], dtype=dtype, device="mps")
    principal_point = torch.tensor([4.0, 3.0], dtype=dtype, device="mps")
    print("mps", dtype, undistort_points_kannala_brandt(principal_point, params).tolist())

# only params.dtype matters: the body casts the points to params.dtype and casts back at the end
params = torch.tensor([100.0, 100.0, 4.0, 3.0, 0.1, 0.01, 0.001, 0.0001], dtype=torch.float32)
print("f16 points / f32 params", undistort_points_kannala_brandt(torch.tensor([4.0, 3.0], dtype=torch.float16), params).tolist())

# the affine sibling, which has no hardcoded guards
for dtype in (torch.float64, torch.float32, torch.bfloat16, torch.float16):
    affine = torch.tensor([100.0, 100.0, 4.0, 3.0], dtype=dtype)
    print("affine", dtype, undistort_points_affine(torch.tensor([4.0, 3.0], dtype=dtype), affine).tolist())

print(torch.tensor(1e-8, dtype=torch.float16).item(), torch.tensor(1e-16, dtype=torch.float16).item())
```

## Observed

```
torch.float64 [0.0, 0.0]
torch.float32 [0.0, 0.0]
torch.bfloat16 [0.0, 0.0]
torch.float16 [nan, nan]
mps torch.float32 [0.0, 0.0]
mps torch.float16 [nan, nan]
f16 points / f32 params [0.0, 0.0]
affine torch.float64 [0.0, 0.0]
affine torch.float32 [0.0, 0.0]
affine torch.bfloat16 [0.0, 0.0]
affine torch.float16 [0.0, 0.0]
0.0 0.0
```

## Expected

The principal point undistorts to the origin in every supported dtype, as it already does in `float32`, `float64` and `bfloat16`. The two guards should be derived from the working dtype rather than hardcoded — `torch.finfo(dtype)`'s smallest **subnormal** (`torch.finfo(dtype).smallest_normal` is the smallest *normal* and would de-normalize the whole subnormal band, cf. #4162), or a dtype-scaled floor of the same spirit as the existing `1e-16` / `1e-8` in `float32`.

Note that the `1e-12` in the Gauss-Newton step denominator (`step = (thd - rth) / (d_thd + 1e-12)`) underflows in `float16` for the same reason; it is harmless at the principal point only because `d_thd` is `1.0` there, but it is the same class of defect.

## Classification

**Broken output**, dtype-scoped — a well-defined input silently producing `nan` in one dtype only. Idiom 2 (a hardcoded constant that does not inherit the input's dtype or precision).

## Disposition

n/a — single symbol.

## Environment

kornia @ `e8111a61` | torch 2.14.0 | python 3.11 | macOS | cpu and mps.

Found while writing the per-op `Convention:` block for this function in the camera conventions batch (#4262, sub-batch 5b). The block's guard bullet is scoped to the dtypes where the claim holds and points here; the claim is pinned by `test_convention_principal_point_undistorts_to_the_origin` and the current `float16` behavior by `test_wart_principal_point_undistorts_to_nan_in_float16_4308`, both in `tests/geometry/camera/test_distortion.py`. The batch documents and pins this behavior; it does not change it.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by ducha-aiki** [comment]
The same `1e-8` rescale guard has a second effect in `float64`: it caps the round-trip accuracy of `undistort_points_kannala_brandt` at about `1e-8` on the normalized plane, far above the dtype's rounding floor.

`mag = |tan θ| / (rth + 1e-8)` scales every result by `1 - 1e-8 / r` instead of `1`, so the absolute error is `|x| · 1e-8 / r` (at most `1e-8`). Measured on kornia#4312's head `d2fe9507` (torch 2.14.0, cpu), `params = [100, 100, 4, 3, 0.1, 0.01, 0.001, 0.0001]`, round trip `undistort(distort(p))`:

| point | float64, `eps = 1e-8` | float64, `eps = 0` | predicted `|x| · 1e-8 / r` | float32, `eps = 1e-8` |
| --- | --- | --- | --- | --- |
| `(0.5, 0.25)` | 9.55e-09 | 5.6e-17 | 8.9e-09 | 2.98e-08 |
| `(3, 0)` | 2.03e-08 | 0.0 | 1.0e-08 | 1.91e-06 |
| `(0.01, 0)` | 1.00e-08 | 1.7e-18 | 1.0e-08 | 1.12e-08 |

The `eps = 0` column is the function re-executed with only that constant changed; the ten Gauss-Newton steps themselves converge to the float64 floor. `float32` is unaffected because its rounding floor sits above the bias. So a dtype-aware `eps` alone does not settle this issue: the guard should stop being an additive term in the denominator (a `where` on `rth > 0`, or a relative epsilon), which fixes the float16 `nan` and the float64 bias together. #4312 now states the float64 floor in the Convention block and pins it as a wart (`test_wart_rescale_guard_caps_float64_round_trip_4308`), to be deleted with this issue.

Posted on behalf of @ducha-aiki by Claude (Claude Fable 5.1).
